### PR TITLE
add GenerationChanged predicate filter to all resources

### PR DESCRIPTION
**Pull request by cnvergence**
**What type of PR is this?**

**What this PR does / why we need it**:
Per review of https://github.com/envoyproxy/gateway/pull/1938#discussion_r1358956623, it would be valuable to add GenerationChanged predicate to all resources, except Service and Deployment.
This skip events that have no change in the object Spec and only the metadata and/or status fields are changed.

### CODE REVIEW

File: internal/provider/kubernetes/controller.go [hunk 1]
```diff
@@ -1317,7 +1319,10 @@ func (r *gatewayAPIReconciler) watchResources(ctx context.Context, mgr manager.M
 	}
 
 	// Watch Gateway CRUDs and reconcile affected GatewayClass.
-	gPredicates := []predicate.Predicate{predicate.NewPredicateFuncs(r.validateGatewayForReconcile)}
+	gPredicates := []predicate.Predicate{
+		predicate.NewPredicateFuncs(r.validateGatewayForReconcile),
+		predicate.GenerationChangedPredicate{},
```
[the remark this answers is not in the document]
Comment (cnvergence) [reply internal/provider/kubernetes/controller.go:1324]: I had no time to investigate yesterday, it seems that the problem is with deployment predicates, as we are checking for 
AvailableReplicas in its status :)

[the remark this answers is not in the document]
Comment (arkodg) [reply internal/provider/kubernetes/controller.go:1323]: ah good catch !
 this is also true for `Service` where we the `LoadBalancerIp` lives In the `status`.
 
 Here's my suggestion -
 * lets rm this new predicate for Service and Deployment
 * lets move the predicate back on top for the rest

[the remark this answers is not in the document]
Comment (cnvergence) [reply internal/provider/kubernetes/controller.go:1323]: alright, I will also add service then

**Review by arkodg — APPROVED** [review]
LGTM thanks !

File: internal/provider/kubernetes/controller.go [hunk 2]
```diff
@@ -1435,6 +1440,7 @@ func (r *gatewayAPIReconciler) watchResources(ctx context.Context, mgr manager.M
 		if err := c.Watch(
 			source.Kind(mgr.GetCache(), &mcsapi.ServiceImport{}),
 			handler.EnqueueRequestsFromMapFunc(r.enqueueClass),
+			predicate.GenerationChangedPredicate{},
```
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1443]: I think you missed `EndpointSlice`, can you please also add the predicate for `EndpointSlice` ?

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zirain — APPROVED** [review]

**Pull request merged**